source/utils/exceptions.cpp

```diff
@@ -146,16 +125,23 @@ no_visible_worksheets::no_visible_worksheets()
 {
 }
 
-no_visible_worksheets::~no_visible_worksheets()
+invalid_password::invalid_password(const std::string &message)
+    : exception("invalid password: " + message)
 {
 }
 
 unsupported::unsupported(const std::string &message)
-    : exception(message)
+    : exception("unsupported: " + message)
+{
+}
+
+encoding_error::encoding_error(const std::string &message)
+    : exception("encoding error: " + message)
 {
 }
 
-unsupported::~unsupported()
+bad_variant_access::bad_variant_access(int expected_type, int actual_type)
+    : exception("bad variant access: expected type " + std::to_string(expected_type) + " but got type " + std::to_string(actual_type))
```

Comment:
It may be better to print the official name (i.e. i4, date, ...) instead of a random number we assigned to it.